inference-explainers/inference_api.md

```diff
@@ -0,0 +1,252 @@
+
+## Introduction
+
+This document provides a comprehensive overview of the Bidding and Auction (B&A) Inference service API with externally supplied models ([B&A inference overview][1]), including supported machine learning (ML) frameworks, model formats, model management processes, and the JavaScript API for executing inference tasks.
+
+## Supported Machine Learning Frameworks and Model Formats
+
+The Tensorflow and PyTorch runtimes are supported, each with a dedicated inference service sidecar binary. Currently, we offer the Tensorflow v2.14.0 sidecar binary and the PyTorch v2.1.1 sidecar binary.
+
+For Tensorflow, the SavedModel format is used while PyTorch uses the TorchScript format. We recommend that ad techs develop and train models in an environment in which the ML runtime version aligns with the intended sidecar.
+
+The maximum size for a single ML model is 2GB.
+
+## Model management
+
+ML models used by the inference service are fetched periodically from a linked cloud storage (e.g., [Google Cloud Storage buckets][2], [Amazon S3 buckets][3]) and made available for serving. Each model is uniquely identified by its path within the cloud storage. To determine the models to load, the inference service looks for a JSON model configuration file in the cloud bucket. Ad techs should maintain a model configuration file in JSON format within the same cloud storage as the models. The B&A service periodically checks the configuration file (according to period configurable by ad techs) for any changes and triggers the loading of new models into the inference service sidecar’s memory as needed.
+
+Ad techs can implement model versioning by structuring model storage paths to include version identifiers. For example, storing models at paths such as “pcvr_v1” and “pcvr_v2” distinguishes between the two versions.
+
+For example:
+
+```json
+{
+  "model_metadata": [
+    {
+      "model_path": "pcvr_v1",
+      "checksum": "dd94b3b08ea19b4240aa5e5f68ff0447b40e37ebdd06cc76fa1a2cf61143a10c",
+      "warm_up_batch_request_json": "..."
+    },
+    {
+      "model_path": "pcvr_v2",
+      "checksum": "...",
+      "warm_up_batch_request_json": "..."
+    }
+  ]
+}
+```
+
+This model configuration file features a top-level array with each entry containing the metadata of the fetched models. The mandatory **model_path** field specifies the path to the model in the cloud storage. If the specified model path points to a directory, it needs to end with a "/" suffix. This suffix indicates that the entire directory will be used for model registration. Without the "/" suffix, an exact path match is expected, and only the specified file will be registered as a model. In the example above, both model paths refer to single model files. If the paths were to refer to directories instead, they would need to be modified to "pcvr_1/" and "pcvr_2/". The optional **checksum** field is the SHA256 checksum of the model represented as a hexadecimal string.
```

Comment:
It's a bit confusing.

Can we give two concrete examples?

For PyTorch, can we use the path like pcvr_v1/model.pt to clearly show that model.pt is a single model file that we're using?

For TensorFlow, a model consists of multiple files and sub-directories, so we can show the example of "pcvr_v1/" where saved_model.pb and other files are in the directory.

Reply:
Done